perception/autoware_tensorrt_bevdet/src/bevdet_node.cpp

```diff
@@ -0,0 +1,266 @@
+    sampleData_.param.cams2ego_trans, onnx_file_, engine_file_);
+
+  RCLCPP_INFO_STREAM(this->get_logger(), "Successful create bevdet!");
+
+  CHECK_CUDA(cudaMalloc(
+    reinterpret_cast<void **>(&imgs_dev_), img_N_ * 3 * img_w_ * img_h_ * sizeof(uchar)));
+
+  pub_cloud_ = this->create_publisher<sensor_msgs::msg::PointCloud2>(
+    "~/output/painting_cloud", rclcpp::SensorDataQoS());
+  pub_boxes_ = this->create_publisher<autoware_perception_msgs::msg::DetectedObjects>(
+    "~/output/boxes", rclcpp::QoS{1});
+
+  sub_cloud_.subscribe(
+    this, "~/input/topic_cloud",
+    rclcpp::QoS{1}.get_rmw_qos_profile());  // rclcpp::QoS{1}.get_rmw_qos_profile()
+  sub_f_img_.subscribe(
+    this, "~/input/topic_img_f",
+    rclcpp::QoS{1}.get_rmw_qos_profile());  // rmw_qos_profile_sensor_data
+  sub_b_img_.subscribe(this, "~/input/topic_img_b", rclcpp::QoS{1}.get_rmw_qos_profile());
+
+  sub_fl_img_.subscribe(this, "~/input/topic_img_fl", rclcpp::QoS{1}.get_rmw_qos_profile());
+  sub_fr_img_.subscribe(this, "~/input/topic_img_fr", rclcpp::QoS{1}.get_rmw_qos_profile());
+
+  sub_bl_img_.subscribe(this, "~/input/topic_img_bl", rclcpp::QoS{1}.get_rmw_qos_profile());
+  sub_br_img_.subscribe(this, "~/input/topic_img_br", rclcpp::QoS{1}.get_rmw_qos_profile());
+
+  sync_ = std::make_shared<Sync>(
+    MySyncPolicy(10), sub_cloud_, sub_fl_img_, sub_f_img_, sub_fr_img_, sub_bl_img_, sub_b_img_,
+    sub_br_img_);
+
+  sync_->registerCallback(
+    std::bind(&TRTBEVDetNode::callback, this, _1, _2, _3, _4, _5, _6, _7));  // 绑定回调函数
+}
+
+void TRTBEVDetNode::callback(
+  const sensor_msgs::msg::PointCloud2::ConstSharedPtr & msg_cloud,
+  const sensor_msgs::msg::Image::ConstSharedPtr & msg_fl_img,
+  const sensor_msgs::msg::Image::ConstSharedPtr & msg_f_img,
+  const sensor_msgs::msg::Image::ConstSharedPtr & msg_fr_img,
+  const sensor_msgs::msg::Image::ConstSharedPtr & msg_bl_img,
+  const sensor_msgs::msg::Image::ConstSharedPtr & msg_b_img,
+  const sensor_msgs::msg::Image::ConstSharedPtr & msg_br_img)
+{
+  pcl::PointCloud<PointT>::Ptr cloud(new pcl::PointCloud<PointT>);
+
+  pcl::fromROSMsg(*msg_cloud, *cloud);
+
+  cv::Mat img_fl, img_f, img_fr, img_bl, img_b, img_br;
+  std::vector<cv::Mat> imgs;
+  img_fl = cv_bridge::toCvShare(msg_fl_img, "bgr8")->image;
+  img_f = cv_bridge::toCvShare(msg_f_img, "bgr8")->image;
+  img_fr = cv_bridge::toCvShare(msg_fr_img, "bgr8")->image;
+  img_bl = cv_bridge::toCvShare(msg_bl_img, "bgr8")->image;
+  img_b = cv_bridge::toCvShare(msg_b_img, "bgr8")->image;
+  img_br = cv_bridge::toCvShare(msg_br_img, "bgr8")->image;
+
+  imgs.emplace_back(img_fl);
+  imgs.emplace_back(img_f);
+  imgs.emplace_back(img_fr);
+  imgs.emplace_back(img_bl);
+  imgs.emplace_back(img_b);
+  imgs.emplace_back(img_br);
+
+  std::vector<std::vector<char>> imgs_data;
+  // Mat -> Vetor<char>
+  cvImgToArr(imgs, imgs_data);
+
+  // cpu imgs_data -> gpu imgs_dev
+  decode_cpu(imgs_data, imgs_dev_, img_w_, img_h_);
```

Comment:
cvImgToArr & decode_cpu essentially encodes the raw image into jpeg then decode them back to raw.

I understand that in order to directly utilize the code from bevdet-tensorrt-cpp, which reads from encoded images in the file system, then decode the raw jpeg data and run inference.

In the code here, we compress the raw image first to align with the code base.

I understand it is easier to work like that, but I would like to point out that encoding/decoding six camera streams in real-time can be significantly performance bottleneck, especially on ARM Jetson machines.

Reply:
For convenience, we directly use the `bevdet-tensorrt-cpp` code baseline, where the input module of the model is the `preprocess_plugin`, in order to match the input, we decode the image as `uchar *`. But I don't think this step is the bottleneck of BEV model. The real bottleneck comes from extracting features from six images simultaneously.